Create new pattern: Group Support (#527)
gruetter/innerSourceTest: patterns/2-structured/support-as-a-group.md

## Title

Support as a Group

## Patlet

> What happens when the team or individual that was supporting an InnerSource project is no longer able to do so?

Reviewer: ```suggestion
What happens if a team or individual no longer supports an InnerSource project?
```

> Form a group of interested individuals from anywhere in the company to keep the project going.

Reviewer: ```suggestion
Keep the project alive by forming a group of interested individuals.
```

## Problem

> * A popular InnerSource project is orphaned with no one to support it.

Reviewer: Orphaned implies no support.

Alternatively: A popular InnerSource project is unsupported.

```suggestion
* A popular InnerSource project is orphaned.
```

What will happen next?
Will new teams shy away from using it?
Will the project stagnate and linger until its users eventually are forced to move on to something else?
What a shame if that were to happen to a perfectly good and useful project!

## Context

* Popular InnerSource project.
* No one is actively supporting it.

## Forces

* No one is assigned by their day job to work on it.
* Everyone is busy.
* High cost to migrate off the project.

## Solutions

Call for interested volunteers from anywhere in the company to form a group of [Trusted Committer]s to support the project.
You may need to reach out to specific individuals based on commit or usage history.
It is important that there are enough so that the burden on each can be reasonably small.

When forming, this group should identify or create:

* A chat channel where questions and discussion about the project should happen.
* Clear contributing instructions.
* Link to both of the above in the project README.
* A chat channel just for the [Trusted Committer]s.

The group should do its best to manage these aspects of the project:

> * **Maintenance**.  If the project is flat-out broken for the standard use case, then fix it.

Reviewer: It seems like we're defining how to maintain an InnerSource project in this Patlet.

Should we create one specifically for that, and provide guidance on how to convert a production project to InnerSource here?

Reviewer: I didn't see a Patlet for that in the repo, but I may be missing something.

Reviewer 2: I think you meant pattern, rather than Patlet above? (I know, its confusing!)

Author: It would be great!  Like we've been able to link to [Standard Base Documentation](https://patterns.innersourcecommons.org/p/base-documentation) and
[Communication Tooling](https://patterns.innersourcecommons.org/p/communication-tooling) then we could just link to some pattern that describes the minimum viable support needed for InnerSource.

Author: It would be great if we had it, but I don't think we should gate on it, either.

## Title

Support as a Group

## Patlet

> What happens when the team or individual that was supporting an InnerSource project is no longer able to do so?
> Form a group of interested individuals from anywhere in the company to keep the project going.

Reviewer 2: In the Patlet we are looking for problem + solution in two sentences. I would try to focus the 1st sentence more on what the core fo the problem seems to be, which is that the company is not supporting the maintenance of this project.

```suggestion
A heavily used InnerSource project is loosing its maintainers and no new maintainer is assigned by the company.
By forming a group of interested individuals from anywhere in the company you are able to keep the project going.
```

Reviewer 2: New Patlet is now:
> What happens if a team or individual no longer supports an InnerSource project?
Keep the project alive by forming a group of interested individuals.

## Problem

* A popular InnerSource project is orphaned with no one to support it.
* There isn't an obvious home for it to land.

## Story

A library of UI widgets is used by over 50 projects throughout the company.
The funding for the team that owns the library runs out and the team disbands.
At first, no one notices, but after a while whenever someone asks "who owns it" there isn't an answer.
What will happen next?
Will new teams shy away from using it?
Will the project stagnate and linger until its users eventually are forced to move on to something else?
What a shame if that were to happen to a perfectly good and useful project!

## Context

* Popular InnerSource project.

> * No one is actively supporting it.

Reviewer 2: ```suggestion
* No one is actively supporting it.
* The company cannot assign a new maintainer to the project because ...
```

Suggesting to add the reason for not assigning a new maintainer here, or maybe a list of possible reasons.

This bit got me confused, as it didn't seem logical to me.
The library has a lot of users (50 projects).
The migration cost away from this library is high.
50 x high migration cost = super high cost :)

So why wouldn't the company assign a new maintainer?
Wouldn't it be in their best interest?

It is entirely possible that I am misunderstanding the scenario/context here.
And if yes that's great, so that we can clarify this :)

Author: `The company cannot assign a team to support.`

What are the reasons?  Could be varied.  The point is we are not getting a team for it.

Reviewer 2: I see. Still confused by this, I have to admit :)

Author: The company is probably short-staffed in general.

## Forces

* No one is assigned by their day job to work on it.
* Everyone is busy.

> * High cost to migrate off the project.

Reviewer 2: This is high costs for the teams that are using this library, right?

Author: Yes!

## Solutions

Call for interested volunteers from anywhere in the company to form a group of [Trusted Committer]s to support the project.
You may need to reach out to specific individuals based on commit or usage history.
It is important that there are enough so that the burden on each can be reasonably small.

When forming, this group should identify or create:

* A chat channel where questions and discussion about the project should happen.
* Clear contributing instructions.
* Link to both of the above in the project README.

> * A chat channel just for the [Trusted Committer]s.

Reviewer 2: We have some other patterns that are cover some of these points already.

It is certainly fine to repeat key elements in this pattern.
By linking to other existing patterns we can let the readers know that they can go into further depth on those topics.

Some relevant patterns that we could link to are:
- https://patterns.innersourcecommons.org/p/base-documentation
- https://patterns.innersourcecommons.org/p/communication-tooling

Author: Good point!  Linked instead.

The group should do its best to manage these aspects of the project:

* **Maintenance**.  If the project is flat-out broken for the standard use case, then fix it.
Keep the project up-to-date as the dependencies and frameworks it uses continue to evolve.
* **Onboarding**.  If someone has a question about how to use the project, make sure they get a reasonable answer.
* **Updates**.  If someone wants to add new feature to the project, give them the design and technical support necessary for them to build it so that it both works for them and is a good addition to the project.

> Review incoming pull requests in a timely manner.

Reviewer 2: I got the sense from this description that the Support Group won't typically implement requested features, will they? Instead they would try the help the requester to implement it themselves?

Maybe add something like this as well. i.e. what the Support Group won't do, or will unlikely be able to do, given their voluntary nature and "best effort" approach.

Author: `The group is not expected to implement any new functionality for others.`

Since this group is comprised of volunteers, it is important to communicate that support is "best effort" only.
Accordingly this model of support is not well-suited for run-time critical, production projects like live APIs.

> It is better suited for projects that are consumed at build-time like libraries/packages/modules.

Reviewer 2: This limitation seems so convincing, that I would even add something about this to the Context.

Maybe:
* the project is a library/package/module that is integrated into other software components at build-time (see Solution for reasoning)

Author: I've having trouble with putting something in the `Context` that makes sense without having to first understand the `Solution`.  Thinking of leaving things as they are?

Reviewer 2: I would find it relevant to put something into the Context, as it is a key limitation, that readers should compare to their own Context.

i.e. knowing that the pattern is meant to be applied to build-time dependencies should be highlighted early.

What were your concern with the proposal above?

Author: I think it's difficult to understand why a build-time dependency is a part of the context without first reading the solution.

Author: I added a line about it in the **Context**.

## Resulting Context

> * There is some "best-effort" support for the InnerSource project.

Reviewer 2: How long is the project expected to survive in this form of support?

Is the hope that eventually a new maintainer will be identified by the company?
If yes, then this could be added here.

Author: It could go on for an indefinite amount of time like this.
Maybe they find themselves one day in a context that demands another pattern like [Core Team](https://patterns.innersourcecommons.org/p/core-team).
But who's to say?  Should we outline these future possibilities or just stick to a succinct declaration of this situation?

Author: Is there a way to link to related patterns?

Reviewer 2: > Is there a way to link to related patterns?

You can link to patterns as part of the regular text in any section, just like you already did.

Often times this happens in the sections
a) "Solution" - think "pattern X is already good, but we needed something else for reasons A, B, C"
b) "Resulting Context" - think "the situation has been improved by implementing this pattern. however as a result another issue may arise, which can be addressed by pattern XYZ"

You could also add a new section `## Related Patterns`.

I favor the former approach. You would have to explain anyways how you see the patterns related, and doing so as part of the common sections of the pattern seems most natural.

## Rationale

People generally want to help.
If there is personal outreach for someone to join as a [Trusted Committer], there are generally a number of people that will say "yes".
Feeling part of a group and being given some structure and responsibility generally motivates people to try their best, which many times ends up being enough.

## Known Instances

* WellSky

## Status

> Structured

Reviewer 2: Would you be ok to share this pattern as "Initial" to get started?
That way it won't be published in our online book immediately, and gives us a bit more time to harden the content.

I admit that this is likely a flaw in our maturity process but somehow it makes me nervous to publish a pattern straight from zero into our book.

Would be honestly great to get your thoughts on this topic, as I feel like I am stuck in my own head here.

Reviewer 2: Adding on to this:
We can decide at the very end whether we want to publish this as Initial or Structured.
So don't worry about this too much for now.

Author: OK we'll see at the end.  Reading the definition of [Maturity Levels](https://github.com/InnerSourceCommons/InnerSourcePatterns/blob/main/meta/contributor-handbook.md#maturity-levels) it does say that we need multiple instances before it can be included as `Structured`.

Reviewer 2: Hmm. I am reading this for `Structured`:

> Is validated by at least 1 (one) known instance

For the most part this has meant that patterns with one Known Instance will be moved to Structured. But of course we can make our own choice.

Author: Oh I see.  Only one required for structured.  I think that this is why I submitted this as `Structured` initially - it's something that I am using at WellSky.

Reviewer 2: Yeah. I think we should stick to the definition as outlined in our Maturity Levels.
That means that the pattern is **Structured**, as it has 1 Known Instance, and also meets all other requirements for that level.

I still think that it might be a flaw in our maturity process, as a single org can publish a pattern straight from zero into our book (with themselves as a known instance).

We (as the patterns working group) should possibly review this in the future but for now let's move on and get this PR merged :)

Reviewer 2: FYI I will also post something similar as my review comment, so that this does not get lost in this thread.
